src/mongocxx/options/client_encryption.hpp

```diff
@@ -56,7 +56,7 @@ class MONGOCXX_API client_encryption {
     /// @return
     ///   An optional pointer to the key vault client.
     ///
-    const stdx::optional<client*>& key_vault_client() const;
+    const stdx::optional<mongocxx::client*>& key_vault_client() const;
```

Comment:
After rearaging the includes, the compiler interpreted `client` in this declaration as `mongocxx::v_noabi::options::client`. However, in the definition (`client_encription.cpp:32`) `client` explicitly refers to a `mongocxx::client`. 

I'm not sure why this didn't generate a compiler warning. Still, there's a few code smells here that could've contributed to this: using a raw pointer (as opposed to the std::unique_ptr, etc.), not including what you use (neither `options/client.hpp` nor `mongocxx/client.hpp` is included above) and a forward declaration of `client`.

[evg task](https://evergreen.mongodb.com/task_log_raw/cxx_driver_ubuntu1604_release_compile_and_test_with_shared_libs_patch_dfe361bf672809beba0f6164fafad9b088d55fef_5fc54d51d6d80a31f05432fb_20_11_30_19_51_56/0?type=T#L2119)

Reply:
I'm surprised the ambiguity did not generate a compiler warning either.

Those are good points. I don't know of any rationale behind using a raw pointer vs. an optional of unique pointer. Since the compiler does not need to know the definition of a pointer type, I'd wager that was the original intent behind doing a forward declare instead of including `mongocxx/client.hpp` directly.